pkg/controller/cassandracluster/statefulset.go

```diff
@@ -136,8 +136,18 @@ func statefulSetsAreEqual(sts1, sts2 *appsv1.StatefulSet) bool {
 			return false
 		}
 	*/
+	containersSpecSts1Len := len(sts1.Spec.Template.Spec.Containers)
+	containersSpecSts2Len := len(sts2.Spec.Template.Spec.Containers)
```

Comment:
I propose you create 2 new variables to store `sts1.Spec.Template.Spec.Containers[I]` and `sts2.Spec.Template.Spec.Containers`. Cause it's also used later to access each container in those lists. That would make the code more readable.

Why not just use what we proposed in the past using a diff library ? I use `github.com/r3labs/diff` in a unit test but there might be a better approach.

Reply:
I tried to use the `patch` module from **banzaicloud** : 

```go
	desiredType := reflect.TypeOf(sts2)
	patchResult, err := patch.DefaultPatchMaker.Calculate(sts1, sts2)

	if err != nil {
		log.Error(err, "could not match objects", "kind", desiredType)
	} else if !patchResult.IsEmpty(){
		logrus.WithFields(logrus.Fields{"statefulset": sts1.Name,
			"namespace": sts1.Namespace}).Info("resource diffs",
			"patch", string(patchResult.Patch),
			"current", string(patchResult.Current),
			"modified", string(patchResult.Modified),
			"original", string(patchResult.Original))
		return false
	}
```

But he find systematically a difference in the `volumeClaimTemplate` : 

```json 
{
    "spec": {
        "volumeClaimTemplates": [
            {
                "metadata": {
                    "labels": {
                        "app": "cassandracluster",
                        "cassandracluster": "cassandra-bgl",
                        "cassandraclusters.db.orange.com.dc": "dc3",
                        "cassandraclusters.db.orange.com.rack": "rack1",
                        "cluster": "casskop",
                        "dc-rack": "dc3-rack1"
                    },
                    "name": "data"
                },
                "spec": {
                    "accessModes": [
                        "ReadWriteOnce"
                    ],
                    "resources": {
                        "requests": {
                            "storage": "20Gi"
                        }
                    },
                    "storageClassName": "local-storage"
                },
                "status": {}
            },
            {
                "metadata": {
                    "labels": {
                        "app": "cassandracluster",
                        "cassandracluster": "cassandra-bgl",
                        "cassandraclusters.db.orange.com.dc": "dc3",
                        "cassandraclusters.db.orange.com.rack": "rack1",
                        "cluster": "casskop",
                        "dc-rack": "dc3-rack1"
                    },
                    "name": "gc-logs"
                },
                "spec": {
                    "accessModes": [
                        "ReadWriteOnce"
                    ],
                    "resources": {
                        "requests": {
                            "storage": "10Gi"
                        }
                    },
                    "storageClassName": "local-storage"
                },
                "status": {}
            },
            {
                "metadata": {
                    "labels": {
                        "app": "cassandracluster",
                        "cassandracluster": "cassandra-bgl",
                        "cassandraclusters.db.orange.com.dc": "dc3",
                        "cassandraclusters.db.orange.com.rack": "rack1",
                        "cluster": "casskop",
                        "dc-rack": "dc3-rack1"
                    },
                    "name": "cassandra-logs"
                },
                "spec": {
                    "accessModes": [
                        "ReadWriteOnce"
                    ],
                    "resources": {
                        "requests": {
                            "storage": "10Gi"
                        }
                    },
                    "storageClassName": "local-storage"
                },
                "status": {}
            }
        ]
    }
}
```

So I think that it doesn't work well for now :/